grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/sets/ShortcutFragmentSet.java

```diff
@@ -19,24 +19,126 @@
+ *  <li>There is a {@link RolePlayerFragmentSet} from {@code c} to {@code x}</li>
+ *  <li>If there is a {@link IsaFragmentSet} from {@code c} to {@code C}, then {@code C} must have a
+ *  {@link LabelFragmentSet}</li>
+ *  <li>If there is a {@link IsaFragmentSet} from {@code r} to {@code R} then {@code R} must have no direct
+ *  sub-types</li>
+ * </ol>
+ *
+ * And optionally:
+ *
+ * <ol>
+ *  <li>There is a {@link IsaFragmentSet} from {@code c} to a type with a {@link LabelFragmentSet} and no direct
+ *  sub-types</li>
+ *  <li>There is a {@link IsaFragmentSet} from {@code r} to a type with a {@link LabelFragmentSet}</li>
+ * </ol>
+ *
+ * We assume that {@code c} is otherwise never referred to in the query, since it's a casting.
+ *
+ * When these criteria are met, all the fragments can be replaced with a {@link ShortcutFragmentSet} from {@code r}
+ * to {@code x}, with optionally specified role- and relation-types.
+ *
  * @author Felix Chapman
  */
 class ShortcutFragmentSet extends EquivalentFragmentSet {
 
     ShortcutFragmentSet(
-            Optional<TypeLabel> roleTypeA, VarName rolePlayerA,
-            Optional<TypeLabel> roleTypeB, VarName rolePlayerB, Optional<TypeLabel> relationType) {
+            VarName relation, VarName edge, VarName rolePlayer, Optional<TypeLabel> roleType,
+            Optional<TypeLabel> relationType) {
         super(
-                Fragments.shortcut(relationType, roleTypeA, roleTypeB, rolePlayerA, rolePlayerB),
-                Fragments.shortcut(relationType, roleTypeB, roleTypeA, rolePlayerB, rolePlayerA)
+                Fragments.inShortcut(rolePlayer, edge, relation, roleType, relationType),
+                Fragments.outShortcut(relation, edge, rolePlayer, roleType, relationType)
         );
     }
+
+    static boolean applyShortcutOptimisation(Collection<EquivalentFragmentSet> fragmentSets, GraknGraph graph) {
+        Iterable<CastingFragmentSet> castingFragmentSets =
+                fragmentSetOfType(CastingFragmentSet.class, fragmentSets)::iterator;
+
+        for (CastingFragmentSet castingFragmentSet : castingFragmentSets) {
+            if (attemptOptimiseCasting(fragmentSets, graph, castingFragmentSet)) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    private static boolean attemptOptimiseCasting(Collection<EquivalentFragmentSet> fragmentSets, GraknGraph graph, CastingFragmentSet castingFragmentSet) {
+        VarName relation = castingFragmentSet.relation();
+        VarName casting = castingFragmentSet.casting();
+
+        RolePlayerFragmentSet rolePlayerFragmentSet = findRolePlayerFragmentSet(fragmentSets, casting);
+
+        // Try and get type of relation
+        Optional<IsaFragmentSet> relIsaFragment = fragmentSetOfType(IsaFragmentSet.class, fragmentSets)
+                .filter(isaFragmentSet -> isaFragmentSet.instance().equals(relation))
+                .findAny();
+
+        Optional<TypeLabel> relType = relIsaFragment
```

Comment:
I had to stare at this for a while before I finally understood it and even now I am still not 100% sure.